### DISCUSSION TITLE
Should ValueTask factor the token value into the GetHashCode result?

### DISCUSSION THREAD

**Discussion by IDisposable** [dotnet/runtime#83685] [Ideas] [answered]
It looks like `ValueTask` and `ValueTask<TResult>` both **correctly** use the _value_ of the `_token` value in the [implementation](https://github.com/dotnet/runtime/blob/d260e5023b6f1593fb38534c81ff030805e9ae0f/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs#L155) of the `Equals` method and the `ValueTask<TResult>` [implementation](https://github.com/dotnet/runtime/blob/d260e5023b6f1593fb38534c81ff030805e9ae0f/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/ValueTask.cs#L552).

The value of `_token` is **not** used in the non-generic `GetHashCode()` implementation, which seems suboptimal. In the non-generic version, we just use the `_obj?.GetHashCode()` value (or zero if null). In the generic version, we just use the `_obj` field which will be `null` for any _successful synchronous completion_, which likely happens a lot. 

When `_obj` references a `IValueTaskSource`, the token is actually meaningful (and used to call `GetStatus()` and `GetResult()`, so if multiple `ValueTask` objects have the same `IValueTaskSource` and different `_token` values then they are definitely distinct object (which the `Equals` implementation **does** get right). 

The same comments apply to the generic `ValueTask<TResult>` usages.

The question is:
Do objects of ValueTask get placed in a collection or used in places where `GetHashCode` returning the same value for `ValueTask`/`ValueType<TResult>` objects that have the same `IValueTaskSource` is problematic? Is putting them into a single collision-heavy bucket an issue?

**Comment by teo-tsirpanis** [comment, marked as answer]
Storing value tasks long-term, especially those backed by an `IValueTaskSource` seems like a code smell. Value tasks' purpose is to be awaited immediately. I don't see any actual impact with their implementation of `GetHashCode()`.
